styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorageTransaction.java

```diff
@@ -179,23 +178,25 @@ public WorkflowId store(Workflow workflow) throws IOException {
   @Override
   public WorkflowId storeWorkflowWithNextNaturalTrigger(Workflow workflow, TriggerInstantSpec triggerSpec)
       throws IOException {
-
-    final Key componentKey = DatastoreStorage.componentKey(tx.getDatastore().newKeyFactory(), workflow.componentId());
-    if (tx.get(componentKey) == null) {
-      tx.put(Entity.newBuilder(componentKey).build());
-    }
-
     final String json = OBJECT_MAPPER.writeValueAsString(workflow);
+
     final Key workflowKey = DatastoreStorage.workflowKey(tx.getDatastore().newKeyFactory(), workflow.id());
-    final Optional<Entity> workflowOpt = DatastoreStorage.getOpt(tx, workflowKey);
-    final Builder entity = DatastoreStorage.asBuilderOrNew(workflowOpt, workflowKey)
-        .set(PROPERTY_WORKFLOW_JSON, StringValue.newBuilder(json).setExcludeFromIndexes(true).build())
+    final Key legacyWorkflowKey = DatastoreStorage.legacyWorkflowKey(tx.getDatastore().newKeyFactory(), workflow.id());
+
+    final Optional<Entity> workflowOpt = getWorkflowOpt(tx, workflow.id(), workflowKey, legacyWorkflowKey);
+
+    final Entity workflowEntity = DatastoreStorage.asBuilderOrNew(workflowOpt, workflowKey)
+        .setKey(workflowKey)
+        .set(PROPERTY_WORKFLOW_JSON,
+            StringValue.newBuilder(json).setExcludeFromIndexes(true).build())
         .set(PROPERTY_NEXT_NATURAL_TRIGGER, instantToTimestamp(triggerSpec.instant()))
-        .set(PROPERTY_NEXT_NATURAL_OFFSET_TRIGGER, instantToTimestamp(triggerSpec.offsetInstant()));
-    tx.put(entity.build());
+        .set(PROPERTY_NEXT_NATURAL_OFFSET_TRIGGER, instantToTimestamp(triggerSpec.offsetInstant()))
+        .build();
 
-    return workflow.id();
+    tx.put(workflowEntity);
+    tx.delete(legacyWorkflowKey);
```

Comment:
Only delete if it existed?

Reply:
This operation has no impact if not exist as I remember. Will double check.